File: plugins/harness/skills/harness-install/scripts/install_harness/installer.py
from __future__ import annotations

from dataclasses import dataclass
import os
import shutil
import subprocess

from .advisory import AdvisoryMixin
from .contracts import ContractsMixin
from .hooks import HooksMixin
from .models import InstallerConfig
from .operations import OperationsMixin
from .paths import normalize_requested_target_path, required_selected_path
from .planning import PlanningMixin
from .preview import PreviewMixin


@dataclass(frozen=True, slots=True)
class ActivationCommand:
    """Command used to activate the selected stack hooks."""

    executable: str
    command: tuple[str, ...]
    success_message: str
    failure_label: str


class HarnessInstaller(
    PlanningMixin,
    PreviewMixin,
    OperationsMixin,
    ContractsMixin,
    HooksMixin,
    AdvisoryMixin,
):
    def __init__(self, config: InstallerConfig) -> None:

        super().__init__(config)
        self.target_root = config.target_root.resolve()

    def run(self) -> None:

        from .errors import fail

        if not self.target_root.is_dir():
            fail(f"target root is not a directory: {self.config.target_root}")
        os.chdir(self.target_root)
        if self.config.action == "preview":
            self.preview_install_set()
            return
        if self.config.action == "show":
            selected_path = normalize_requested_target_path(
                required_selected_path(self.config)
            )
            self.show_one_target_path(selected_path)
            return
        if self.config.action == "only":
            selected_path = normalize_requested_target_path(
                required_selected_path(self.config)
            )
            self.install_one_target_path(selected_path)
            self.print_summary(selected_path)
            self.runtime_advisory_for_mode()
            return
        self.install_full_plan()
        self.activate_git_hooks()
        self.print_summary()
        self.runtime_advisory_for_mode()

    def activate_git_hooks(self) -> None:

        mode = self.config.mode
        if mode == "gradle":
            print(
                "activate git hooks: Gradle plugin creates hooks on first build",
            )
        elif mode in {"maven", "shell"}:
            self.run_activation_command(
                ActivationCommand(
                    executable="git",
                    command=("git", "config", "core.hooksPath", ".githooks/"),
                    success_message="activate git hooks: git config core.hooksPath .githooks/",
                    failure_label="git config core.hooksPath",
                )
            )
        elif mode == "uv":
            self.run_activation_command(
                ActivationCommand(
                    executable="uv",
                    command=("uv", "run", "pre-commit", "install"),
                    success_message="activate git hooks: uv run pre-commit install",
                    failure_label="pre-commit install",
                )
            )
        elif mode == "bun":
            self.run_activation_command(
                ActivationCommand(
                    executable="bun",
                    command=("bun", "install"),
                    success_message="activate git hooks: bun install (Husky prepare)",
                    failure_label="bun install",
                )
            )

    def run_activation_command(self, activation: ActivationCommand) -> None:
        """Run one hook activation command when its executable is installed."""

        if shutil.which(activation.executable) is None:
            print(
                f"[warning] {activation.executable} not in PATH; "
                f"skipping {activation.failure_label}"
            )
            return
        result = subprocess.run(
            activation.command,
            capture_output=True,
            text=True,
            check=False,
        )
        if result.returncode == 0:
            print(activation.success_message)
            return
        print(f"[warning] {activation.failure_label} failed: {result.stderr.strip()}")

File: plugins/harness/skills/harness-install/scripts/install-harness/installer.ts
import { existsSync } from "node:fs";
import path from "node:path";

import { validationCommandForMode } from "./commands.js";
import { hookActivationMessage, activateHooks } from "./hook-activation.js";
import { installFullPlan, installOneTargetPath } from "./operations.js";
import { normalizeRequestedTargetPath, requiredSelectedPath } from "./paths.js";
import { previewInstallSet, showOneTargetPath } from "./preview.js";
import { fail } from "./types.js";
import type { InstallerConfig, Mode } from "./types.js";

const executableExists = (executable: string): boolean =>
  Bun.which(executable) !== null;

/** Clear Git state inherited from a caller before operating on an explicit target. */
const isolateTargetGitEnvironment = (): void => {
  for (const key of Object.keys(process.env)) {
    if (key.startsWith("GIT_")) {
      Reflect.deleteProperty(process.env, key);
    }
  }
};

const printSummary = (
  config: InstallerConfig,
  onlySelected: null | string
): void => {
  console.log("");
  console.log(`target: ${process.cwd()}`);
  console.log(`mode: ${config.mode}`);
  console.log(`ci-host: ${config.ciHost}`);
  console.log(`validation command: ${validationCommandForMode(config.mode)}`);
  if (onlySelected !== null) {
    console.log(`selected file: ${onlySelected}`);
  }
};

const runtimeAdvisoryForMode = (mode: Mode, hooksActive: boolean): void => {
  const activationMessage = hookActivationMessage();
  switch (mode) {
    case "gradle": {
      if (!existsSync("./gradlew")) {
        console.error(
          "[advisory] ./gradlew is required before running validation; add or restore the Gradle wrapper in the target repository."
        );
      }
      break;
    }
    case "maven": {
      if (!existsSync("./mvnw")) {
        console.error(
          "[advisory] ./mvnw is required before running validation; add or restore the Maven wrapper in the target repository."
        );
      }
      break;
    }
    case "uv": {
      if (!executableExists("uv")) {
        console.error(
          "[advisory] uv command not found on PATH; install via the official script (`curl -LsSf https://astral.sh/uv/install.sh | sh`) or Homebrew (`brew install uv`) before running validation."
        );
      }
      break;
    }
    case "bun": {
      if (!executableExists("bun")) {
        console.error(
          "[advisory] bun command not found on PATH; install via the official script (`curl -fsSL https://bun.sh/install | bash`) or Homebrew (`brew install oven-sh/bun/bun`) before running validation."
        );
      }
      break;
    }
    case "shell": {
      console.error(
        "[advisory] shellcheck and shfmt are required; install them via your OS package manager (for example, `apt install shellcheck shfmt` on Debian/Ubuntu or `brew install shellcheck shfmt` on macOS) before running validation."
      );
      break;
    }
    default: {
      return fail(
        `unsupported mode (must be gradle|maven|uv|bun|shell): ${mode}`
      );
    }
  }
  console.error(
    hooksActive
      ? `[advisory] Git hooks are active through: ${activationMessage}`
      : `[advisory] To activate Git hooks, run: ${activationMessage}`
  );
};

/** Run the requested installer action from the target repository root. */
export const runInstaller = async (config: InstallerConfig): Promise<void> => {
  const targetRoot = path.resolve(config.targetRoot);
  const targetStat = await Bun.file(targetRoot)
    .stat()
    .catch(() => null);
  if (!existsSync(targetRoot) || targetStat?.isDirectory() !== true) {
    fail(`target root is not a directory: ${config.targetRoot}`);
  }
  process.chdir(targetRoot);
  isolateTargetGitEnvironment();
  switch (config.action) {
    case "preview": {
      await previewInstallSet(config);
      return;
    }
    case "show": {
      await showOneTargetPath(
        config,
        normalizeRequestedTargetPath(requiredSelectedPath(config))
      );
      return;
    }
    case "only": {
      const selectedPath = normalizeRequestedTargetPath(
        requiredSelectedPath(config)
      );
      await installOneTargetPath(config, selectedPath);
      printSummary(config, selectedPath);
      runtimeAdvisoryForMode(config.mode, false);
      return;
    }
    case "install": {
      await installFullPlan(config);
      await activateHooks(config);
      printSummary(config, null);
      runtimeAdvisoryForMode(config.mode, config.activateHooks);
      return;
    }
    default: {
      return fail(`unsupported action: ${config.action}`);
    }
  }
};
